File: tmp/npd_ori/NPDFaceDetector_Train/src/TestDQT.m
function score = TestDQT(tree, x)
% function to test the learned DQT based weak classifier.

n = size(x,1);
score = zeros(n,1,'single');

if isempty(x)
    score(:) = repmat(tree.fit, size(x,1), 1);
else
   score = TestSubTree(tree, x, 0);
end
end


function score = TestSubTree(tree, x, node)

if isempty(x)
    score = [];
    return;
end

n = size(x,1);
score = zeros(n,1,'single');

if node < 0 % leaf node
    score(:) = tree.fit(-node);
else % branch node
    mNode = node + 1;
    
    isLeft = (x(:, mNode) < tree.cutpoint(1, mNode)) | (x(:, mNode) > tree.cutpoint(2, mNode));    
    score(isLeft) = TestSubTree(tree, x(isLeft,:), tree.leftChild(mNode));
    score(~isLeft) = TestSubTree(tree, x(~isLeft,:), tree.rightChild(mNode));
end
end

File: src/npd_train/TestDQT.cpp
#include "common.h"
#include "opencv2/core/core.hpp"
#include <iostream>

using namespace std;

double TestSubTree(Stage& tree,const cv::Mat& x, int node) {
	double score;

	if (tree.feaId.empty()) {
		return 0;
	}

	if (node < 0) {
		score = tree.fit[-node-1];
	} 
	else{
		int mNode = node;

		unsigned char * ptr = x.data;

		bool isLeft = (  ptr[tree.feaId[mNode]] < tree.cutpoint[0][mNode]  || ptr[tree.feaId[mNode]] > tree.cutpoint[1][mNode]);
		
		if (isLeft) {
			score = TestSubTree(tree, x, tree.leftChild[mNode]);
		}
		else {
			score = TestSubTree(tree, x, tree.rightChild[mNode]);
		}
	}

	return score;

}


double TestDQT(Stage& tree, const cv::Mat& x) {
	
	double score;

	if (tree.feaId.empty()) {
			score = tree.fit[0];
	}
	else {
		score = TestSubTree(tree, x, 0);
	}

	return score;
}
